### PR TITLE
Refactor `prepare_model_for_kbit_training`

### PR DESCRIPTION
**Pull request by josephenguehard**
There seems to be 2 issues in `prepare_model_for_kbit_training`:
- `gradient_checkpointing_enable` can only be used if `loaded_in_kbit` is True. This PR decouples these two parameters.
- By default, the base model is frozen. This PR adds the option not to do so, with the arg `freeze_base_model`.

### CODE REVIEW

File: src/peft/utils/other.py [hunk 1]
```diff
@@ -42,19 +42,22 @@ def prepare_model_for_kbit_training(model, use_gradient_checkpointing=True):
...
             The loaded model from `transformers`
+        freeze_base_model (bool): Whether to freeze the base model or not.
+        use_gradient_checkpointing (bool): Whether to enable gradient checkpointing or not.
     """
     loaded_in_kbit = getattr(model, "is_loaded_in_8bit", False) or getattr(model, "is_loaded_in_4bit", False)
 
-    for name, param in model.named_parameters():
-        # freeze base model's layers
-        param.requires_grad = False
+    if freeze_base_model:
```
Comment (pacman100) [inline src/peft/utils/other.py:50]: Want to understand the use case for having this parameter.

**Review by pacman100 — COMMENTED** [review]
Hello @josephenguehard, Thank you for the PR, please find the comments.

File: src/peft/utils/other.py [hunk 2]
```diff
@@ -42,19 +42,22 @@ def prepare_model_for_kbit_training(model, use_gradient_checkpointing=True):
...
+        for name, param in model.named_parameters():
+            # freeze base model's layers
+            param.requires_grad = False
 
     # cast all non INT8 parameters to fp32
     for param in model.parameters():
         if (param.dtype == torch.float16) or (param.dtype == torch.bfloat16):
             param.data = param.data.to(torch.float32)
 
-    if loaded_in_kbit and use_gradient_checkpointing:
```
[the remark this answers is not in the document]
Comment (pacman100) [reply src/peft/utils/other.py:57]: PR #404 does this for cases when gradient ckpt is enabled for non kbit models

**Comment by josephenguehard** [comment]
Hi Sourab,

Thanks for the review, I will close this PR for now, it seems that it is actually not useful for the PEFT library.
